### PR TITLE
Add card functions

### PR DESCRIPTION
**Pull request by benadida**
add backend that talks to the smart card and frontend that queries the backend.

### CODE REVIEW

File: src/pages/ActivationPage.tsx [hunk 1]
```diff
@@ -63,6 +66,23 @@ const StartPage = (props: RouteComponentProps) => {
     event.target.focus()
   }
 
+  // TODO: Mock this timer
+  /* istanbul ignore next */
```
Comment (beausmith) [inline src/pages/ActivationPage.tsx]: We previously agreed that anytime we were to use `istanbul ignore next` that we would document why and/or where the code is tested. eg: `/* istanbul ignore next: Tested by Cypress */`

File: src/pages/ActivationPage.tsx [hunk 2]
```diff
@@ -63,6 +66,23 @@ const StartPage = (props: RouteComponentProps) => {
     event.target.focus()
   }
 
+  // TODO: Mock this timer
```
[the remark this answers is not in the document]
Comment (benadida) [reply src/pages/ActivationPage.tsx]: good callout!

File: src/pages/ActivationPage.test.tsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import React from 'react'
+import { Route } from 'react-router-dom'
+
+import { render } from '../../test/testUtils'
+
+import ActivationPage from './ActivationPage'
+
+// https://www.npmjs.com/package/jest-fetch-mock
```
Comment (beausmith) [inline src/pages/ActivationPage.test.tsx]: nit: delete comment.

File: src/pages/ActivationPage.test.tsx [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import fetch, { FetchMock, GlobalWithFetchMock } from 'jest-fetch-mock'
+const customGlobal: GlobalWithFetchMock = global as GlobalWithFetchMock
```
Comment (beausmith) [inline src/pages/ActivationPage.test.tsx]: This line can simply be: `const customGlobal = global as GlobalWithFetchMock`

File: src/pages/ActivationPage.test.tsx [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+customGlobal.fetch = fetch as FetchMock
+customGlobal.fetchMock = customGlobal.fetch as FetchMock
```
Comment (beausmith) [inline src/pages/ActivationPage.test.tsx]: This line can simply be: `customGlobal.fetchMock = customGlobal.fetch`

File: src/pages/ActivationPage.test.tsx [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    [JSON.stringify({}), { status: 200 }],
+    [JSON.stringify({ card: 'VX.precinct-21.5R' }), { status: 200 }],
+    ['', { status: 500 }]
+  )
+
+  render(<Route path="/" component={ActivationPage} />, {
+    route: '/',
+  })
+
+  expect(setInterval).toHaveBeenCalledTimes(1)
```
Comment (beausmith) [inline src/pages/ActivationPage.test.tsx]: Let's avoid naked globals. Please change to `expect(window.setInterval).toHaveBeenCalledTimes(1)`

File: src/pages/ActivationPage.test.tsx [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+  expect(fetchMock.mock.calls.length).toEqual(3)
+  expect(fetchMock.mock.calls).toEqual([
+    ['/card/read'],
+    ['/card/read'],
+    ['/card/read'],
+  ])
+
+  // TODO: not clear why this one is failing
+  // expect(clearInterval).toHaveBeenCalledTimes(2)
```
Comment (beausmith) [inline src/pages/ActivationPage.test.tsx]: issue link?

File: src/pages/ActivationPage.test.tsx [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    route: '/',
+  })
+
+  expect(setInterval).toHaveBeenCalledTimes(1)
+
+  jest.advanceTimersByTime(1000)
+
+  jest.advanceTimersByTime(1000)
+
+  // TODO: check that setActivationCode has been called
```
Comment (beausmith) [inline src/pages/ActivationPage.test.tsx]: Fix: `decodeActivationCode()` calls `setBallot()` which calls `activateBallot()` which is a context method which you can place in the `render()` method's second argument… and check to see if it has been called.

File: src/pages/ActivationPage.test.tsx (hunk 7 above)
Comment (beausmith) [reply src/pages/ActivationPage.test.tsx]: Seem like this should work. Is `clearInterval` defined? Did you try `window.clearInterval`?

File: src/pages/ActivationPage.test.tsx [hunk 9]
```diff
@@ -0,0 +1,48 @@
+import React from 'react'
```
Comment (beausmith) [inline src/pages/ActivationPage.test.tsx]: Thank you for creating this file! Not sure how it was overlooked before. 🤦‍♂

File: src/pages/ActivationPage.tsx [hunk 10]
```diff
@@ -77,6 +79,23 @@ const StartPage = (props: RouteComponentProps) => {
...
+      .then(resultJSON => {
+        const cardState = resultJSON.card
+        if (!!cardState) {
+          clearInterval(checkCard)
+          setActivationCode(cardState)
+          decodeActivationCode()
+        }
+      })
+      .catch(() => {
+        // if it's an error, aggressively assume there's no backend and stop hammering
```
Comment (beausmith) [inline src/pages/ActivationPage.tsx]: ❤️

File: src/pages/ActivationPage.tsx [hunk 11]
```diff
@@ -77,6 +79,23 @@ const StartPage = (props: RouteComponentProps) => {
...
+        clearInterval(checkCard)
```
Comment (beausmith) [inline src/pages/ActivationPage.tsx]: Let's avoid naked globals. Please change to `window.clearInterval(checkCard)`

File: src/pages/ActivationPage.tsx [hunk 12]
```diff
@@ -77,6 +79,23 @@ const StartPage = (props: RouteComponentProps) => {
     event.target.focus()
   }
 
+  checkCard = window.setInterval(() => {
+    fetch('/card/read')
+      .then(result => result.json())
+      .then(resultJSON => {
+        const cardState = resultJSON.card
+        if (!!cardState) {
+          clearInterval(checkCard)
```
Comment (beausmith) [inline src/pages/ActivationPage.tsx]: Let's avoid naked globals. Please change to `window.clearInterval(checkCard)`

File: src/pages/ActivationPage.test.tsx (hunk 3 above)
Comment (benadida) [reply src/pages/ActivationPage.test.tsx]: the comment was to explain the approach, which isn't obvious, I think.

File: src/pages/ActivationPage.test.tsx (hunk 8 above)
Comment (benadida) [reply src/pages/ActivationPage.test.tsx]: 👍 will try that

File: src/pages/ActivationPage.test.tsx (hunk 7 above)
Comment (benadida) [reply src/pages/ActivationPage.test.tsx]: yep, doesn't work either. I'll update the commented call to make sure we don't keep the bad global, but I'm confused as to why this fails.

**Pull request merged**